### PR TITLE
#1130 New StocktakeManagePage - #4

### LINKED ISSUE #1130: Refactor StocktakeManagePage to use new DataTable
## Is your feature request related to a problem? Please describe.
Need to use the new `DataTable` in CustomerInvoicesPage.

## Describe the solution you'd like

- A more or less complete refactor, using the first refactored page `StocktakeManagePage` as a guide.
- This will require using hooks and the new DataTable `component`.

## Describe alternatives you've considered

N/A

## Additional context

This is a placeholder issue. Please investigate and compare the page with `CustomerInvoicePage`. Each page being refactored may have their own quirks. The tissue is encouraged to describe any additional problems and differences with the particular page here, to find a solution which encourages further extensibility and code-reuse with other pages, if possible.

### PR DESCRIPTION
**Pull request by joshxg**
#### NOTE: BRANCHES FROM #1188 

Fixes #1130 

## Change summary

- Refactors `StocktakeManagePage`

## Testing

SEE #1043 

### Related areas to think about
N/A

### CODE REVIEW

File: src/database/DataTypes/Stocktake.js [hunk 1]
```diff
@@ -88,6 +88,14 @@ export class Stocktake extends Realm.Object {
     });
   }
 
+  get itemsInStocktake() {
```
Comment (joshxg) [inline src/database/DataTypes/Stocktake.js]: doc string

File: src/pages/StocktakeManagePage.js [hunk 2]
```diff
@@ -1,290 +1,201 @@
...
-      this.setState(
+  });
+
+  const Toggle = () => (
+    <ToggleBar
+      style={globalStyles.toggleBar}
+      textOffStyle={globalStyles.toggleText}
+      textOnStyle={globalStyles.toggleTextSelected}
+      toggleOffStyle={globalStyles.toggleOption}
+      toggleOnStyle={globalStyles.toggleOptionSelected}
```
Comment (joshxg) [inline src/pages/StocktakeManagePage.js]: these styles are defaults and not needed to pass through

File: src/pages/StocktakeManagePage.js [hunk 3]
```diff
@@ -1,290 +1,201 @@
...
+      toggles={[
         {
-          selection: selected,
-          stocktakeName: stocktake.name,
+          text: buttonStrings.hide_stockouts,
+          onPress: () => (showAll ? dispatch(hideStockOut()) : dispatch(showStockOut())),
```
Comment (joshxg) [inline src/pages/StocktakeManagePage.js]: this is confusing - should dispatch a `toggleStockOut` and let the action creator determine which action to return to dispatch

File: src/pages/StocktakeManagePage.js [hunk 4]
```diff
@@ -1,290 +1,201 @@
...
+          isOn: !showAll,
         },
-        this.refreshData
+        {
+          text: buttonStrings.all_items_selected,
+          onPress: () => (allSelected ? dispatch(deselectAll()) : dispatch(selectAll())),
```
Comment (joshxg) [inline src/pages/StocktakeManagePage.js]: same as above

File: src/pages/StocktakeManagePage.js [hunk 5]
```diff
@@ -1,290 +1,201 @@
...
+      isAscending={isAscending}
+      sortBy={sortBy}
+    />
+  );
+
+  const confirmStocktake = () => {
+    const itemIds = Array.from(dataState.keys()).filter(id => id);
+
+    const updateExistingStocktake = () => reduxDispatch(addItemsToStocktake(stocktake, itemIds));
+    const createNewStocktake = () =>
```
Comment (joshxg) [inline src/pages/StocktakeManagePage.js]: no reason to wrap these in functions here other than confusion

File: src/pages/dataTableUtilities/reducer/reducerMethods.js [hunk 6]
```diff
@@ -187,7 +187,22 @@ export const deselectAll = state => {
       });
     }
   }
-  return { ...state, dataState: newDataState, hasSelection: false };
+  return { ...state, dataState: newDataState, hasSelection: false, allSelected: false };
+};
+
+export const selectAll = state => {
+  const { data, dataState, keyExtractor } = state;
+  const newDataState = new Map(dataState);
```
Comment (joshxg) [inline src/pages/dataTableUtilities/reducer/reducerMethods.js]: page only uses `isSelected` in `dataState`, just make a new map

Comment (joshxg) [reply src/pages/dataTableUtilities/reducer/reducerMethods.js]: just use `selectItems` with an array of all `rowKeys` and delete this method

File: src/widgets/modals/BottomTextEditor.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable import/prefer-default-export */
```
Comment (joshxg) [inline src/widgets/modals/BottomTextEditor.js]: just a copy paste from `StocktakeManagePage`, in it's own file

File: src/pages/StocktakeManagePage.js [hunk 8]
```diff
@@ -1,290 +1,201 @@
...
+      const { item, index } = listItem;
+      const rowKey = keyExtractor(item);
+      const { row, alternateRow } = newDataTableStyles;
+
+      return (
+        <DataTableRow
+          rowData={data[index]}
+          rowState={dataState.get(rowKey)}
+          rowKey={rowKey}
+          style={index % 2 === 0 ? alternateRow : row}
```
Comment (mark-prins) [inline src/pages/StocktakeManagePage.js]: as mentioned elsewhere : can push this down into the component

File: src/pages/StocktakeManagePage.js (hunk 2 above)
Comment (mark-prins) [reply src/pages/StocktakeManagePage.js]: nice!

File: src/pages/StocktakeManagePage.js (hunk 3 above)
Comment (mark-prins) [reply src/pages/StocktakeManagePage.js]: yep

**Pull request merged**